core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp

```diff
@@ -45,13 +45,14 @@ struct gain_matrix_updater {
 
         const auto D = trk_state.get_measurement().meas_dim;
         assert(D == 1u || D == 2u);
-        if (D == 1u) {
-            return update<1u, shape_type>(trk_state, bound_params);
-        } else if (D == 2u) {
-            return update<2u, shape_type>(trk_state, bound_params);
+        switch (D) {
+            case 1u:
+                return update<1u, shape_type>(trk_state, bound_params);
+            case 2u:
+                return update<2u, shape_type>(trk_state, bound_params);
+            default:
+                return false;
```

Comment:
I'd rather finish this with a `__builtin_unreachable`.

Reply:
It's not quite unreachable though, is it? :thinking: In a release build the assertion would let this code be reached. At which point we can once again go back to my previous thoughtline.
  - Print an error message;
  - Either stop the execution, or let it continue as it currently does.
    * Letting it continue, with an error communicated back to the caller about failing the update, does not seem that crazy to me. :thinking:

Reply:
Of course it is not _technically_ unreachable but it is semantically unreachable, which is exactly what is modelled by the `assert`. I think it's fair to say that the code reaching this point without `D` being either 1 or 2 can be declared undefined behaviour, which is exactly what marking it as unreachable does.